### DISCUSSION TITLE
RT1170 Port of MicroPython

### DISCUSSION THREAD

**Discussion by Thinbungee** [micropython/micropython#10681] [m.iMX RT / Teensy 4.x]
Does anyone have experience of using MicroPython with the RT1170 ?  I am looking to evaluate it using the NXP eval board but am struggling to get started.  I can transfer the .bin build to the eval board seemingly OK, but havent seen any response on the vitual COM port after reset.  What COM settings are required ?

On another tack can anyone give me any tips on building the code from source using the MCUexpresso system.

**Comment by robert-hh** [comment]
Thanks for the interest. The REPL prompt comes up at the USB1 port (J20), not the J11 you use for loading the firmware. The comm setting do not matter, but you can use 115200 baud, 8 bit, 1 stop bit. For convenience, you can set J38 to the 3-4 position to have the board powered by the J20 USB.

The software is built using the MicroPython environment, not MCUexpresso. It may be possible, but will be a long and rough ride.

**Reply by robert-hh** [reply]
After loading the firmware, you have to push reset or do a power cycle.

**Comment by Thinbungee** [comment]
Thanks for your fast response Robert.  I will investigate The USB1 port and also the building process using the MicroPython environment and maybe come back later with some questions.... :)

**Comment by Thinbungee** [comment]
I have the REPL prompt - many thanks.  Can you recommend an IDE to run on a PC ?

**Reply by robert-hh** [reply]
I hope you uploaded the latest nightly build. It fixes a bug which prevents more than 1MByte for the python heap.
I do not use and IDE, so I cannot recommend one. Many people use Thonny, so it should work with the 1170 board as well. It might need some configuration. For file transfer you can use mpremote or indeed Thonny. Or you can install a ftp server on the board. I have attached slightly modified version of the one from this URL: https://github.com/robert-hh/FTP-Server-for-ESP8266-ESP32-and-PYBD

[wftp.zip](https://github.com/micropython/micropython/files/10662827/wftp.zip)

**Reply by Thinbungee** [reply]
Are you aware of the status of the _thread module on RT1170 port ?  I have been able to do a fair bit of testing with the RT1170 Eval board after your help.  Many thanks !!!  Have been using Thonny so far.

**Reply by robert-hh** [reply]
Threading is not supported. It should be possible to implement it, but it's not done. You may try using asyncio instead.

**Reply by Thinbungee** [reply]
It seems other Micropython ports have the "_thread" module. Do you have any information about why it isnt in RT1170 currently ? Maybe I should try on other hardware then look at making it function on RT1170 ?

**Reply by robert-hh** [reply]
Not all ports have a thread module. And it does not exist because I did not implement it. If you want to implement it, have a look at STM32, files mpthreadport.* and pybthread,*. These should be most of what it needs, in addition to matching changes in Makefile and mpconfigport.h. And testing of course.

**Reply by Thinbungee** [reply]
Thanks for your comments which are valuable to me.  I think I will try and build a better understanding of the software before tackling threads for this RT1170 port.  The important thing is that it is feasible in the future.  I will try and test _thread with other hardware soon.

**Reply by robert-hh** [reply]
I made a first attempt for Threading in the mimxrt port. The branch is here: https://github.com/robert-hh/micropython/tree/mimxrt_thread
First tested on a MIMXRT1020 board, for the only reason that this board is the most convenient one when run with the debugger. I can start several threads and the run concurrently. Further tests are required.
It was a little bit more work than I wrote above, but mostly using the STM32 port as template.

**Reply by Thinbungee** [reply]
That is great !  Currently I havent yet got a working build system.  Can you build an RT1170 image for me to try it too ?

**Reply by robert-hh** [reply]
Will do, once it works with the 1170.

**Reply by robert-hh** [reply]
Here's one for trial: https://github.com/robert-hh/Shared-Stuff/blob/master/firmware_mimxrt1170_thread.bin
To be further tested.

**Reply by Thinbungee** [reply]
I tried the binary you made and it works as far as I have tested it.  Many, many thanks Robert !  It seems there is no method currently for threads to read the status of the other threads ?  Currently am just using globals which is clearly non-atomic.  Can I ask which "debugger" you were referring to when you posted yesterday ?

**Reply by robert-hh** [reply]
I'm using the Segger Ozone debugger with the Segger J-Link EDU adapter. That works fine. Ozone complains on the first boot per day that it is not licensed, but according to their license conditions that is fine for non-commercial use. 
Other debuggers and adapters may work as well, but I did not try it. A 10$ so-called compatible adapter did not work.

**Reply by robert-hh** [reply]
I ran the MP test set for thread with the firmware and all tests passed fine. Test done both with a Teensy board and the MIMXRT1170_EVK board.

**Reply by Thinbungee** [reply]
Great, it's really good to know that the thread tests OK.  I am currently trying to get a build setup working.

**Reply by Thinbungee** [reply]
Hi Robert, any ideas where an error in building the cross compiler might come from ?

cdb@DESKTOP-7A1EQBP:~/micropython/mpy-cross$ make
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
CC ../py/emitnx64.c
In file included from ../py/emitnx64.c:16:
../py/emitnative.c: In function 'emit_native_binary_op':
../py/emitnative.c:2368:38: error: comparison is always true due to limited range of data type [-Werror=type-limits]
 2368 |         } else if (MP_BINARY_OP_LESS <= op && op <= MP_BINARY_OP_NOT_EQUAL) {
      |                                      ^~
cc1: all warnings being treated as errors
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [../py/mkrules.mk:83: build/py/emitnx64.o] Error 1

**Reply by robert-hh** [reply]
If the repository is recent, it may be a compiler issue. Which version of the gcc compiler do you use? I use version 11.3 or 10.3. As far as I recall, there were problems with version 11.2.

**Reply by Thinbungee** [reply]
I am running 10.3.1   The Cross-compiler should be built for the host I read.  Could you let me have a copy of your py/mkenv.mk please for comparison ?

**Comment by Thinbungee** [comment]
Yes, I uploaded the latest nightly build.  Great to have your advice.  I will have a look at both your FTP server and an IDE and consider.

**Reply by robert-hh** [reply]
my mkenv.mk is the one from the Micropython repository. I never change it.

**Reply by robert-hh** [reply]
It looks more like an issue in the default data types, like 16 bit instead of 32 bit integers

**Reply by Thinbungee** [reply]
I updated gcc to 11.3 and get to the link stage now for the cross compiler make:

...
CC main.c
CC gccollect.c
CC ../shared/runtime/gchelper_generic.c
LINK build/mpy-cross
/usr/bin/ld: build/py/nlr.o: relocations in generic ELF (EM: 40)
/usr/bin/ld: build/py/nlr.o: relocations in generic ELF (EM: 40)
/usr/bin/ld: build/py/nlr.o: error adding symbols: file in wrong format
collect2: error: ld returned 1 exit status
make: *** [../py/mkrules.mk:206: build/mpy-cross] Error 1
cdb@DESKTOP-7A1EQBP:~/micropython/mpy-cross$

**Reply by Thinbungee** [reply]
Have got the cross-compiler to build by running on a Raspberry Pi :)

**Reply by robert-hh** [reply]
Quite a few people use a Raspberry Pi as build platform, when Windows or Mac have problems. For just building a RP4 is sufficiently fast. A friend of mine even uses a RP3, which takes a little bit longer.

**Reply by Thinbungee** [reply]
Success ! Many thanks again for your support !  Have (finally) got Micropython to compile and it loads and runs on the RT1170 EVK.  Did you upload the threads changes you made to github ?  Maybe my fork was before your upload but I dont have the _thread module.  Another question is that my build is 475KB which is rather smaller than your 618KB.  Have you seen such large differences ?  Finally where would be the best place to put some hints on how to build for RT1170 ?

**Reply by robert-hh** [reply]
I pushed the thread branch again, and have built it again and uploaded to the board. It works and shows the _thread module. And the size is > 600k. So if you build is smaller, then there might be a problem. What is the MicroPython version shown after pushing Ctrl-B in REPL?
```
   text	   data	    bss	    dec	    hex	filename
 625352	    604	 100236	 726192	  b14b0	build-MIMXRT1170_EVK/firmware.elf
```

**Reply by Thinbungee** [reply]
Which fork are you pushing to ?  I dont fully understand github yet so maybe am doing something wrong.
>>> print (sys.version)
3.4.0; MicroPython 41ed01f13 on 2023-02-14

**Reply by robert-hh** [reply]
It is this fork: https://github.com/robert-hh/micropython
and the thread branch is mimxrt_thread
Don't forget to update the submodules with `make submodules`.
You compiled the master branch. You have to checkout the mimxrt_thread branch;
`git checkout mimxrt_thread`
Before compiling again, better clean out the old compile with `make BOARD=MIMXRT1170_EVK clean`.

**Reply by Thinbungee** [reply]
I've set your fork and checked out the mimxrt_thread.  It now builds and runs my threading program OK  :)  Thanks again for your support.  There is still the mystery of the smaller code size.  Adding the threading increased my .bin file size marginally to 478KB, which is slightly more than the standard build at 475KB.  The threading adds about 5% to a simple speed test time, although I noticed your thread build adds about 29% to the same test time.  Is it possible you are running with different compiler optimizations ?

**Reply by robert-hh** [reply]
I notice the drop in performance as well, but I could not spot the reason.It must be somewhere in the build system, since the repositories are identical.  At least it's good that the official builds work well.

**Reply by robert-hh** [reply]
Found it. It was in my GNUmakefile, where DEBUG=1 is set for the MIMXRT1170 port. There is a small increase since I usually have addiotional frozen bytecode added to flash, like upysh and the pye editor.

**Reply by Thinbungee** [reply]
Nice to have the mystery resolved !

**Reply by Thinbungee** [reply]
Hi Robert, I recall you were using Segger Ozone with J-Link EDU and I presume the MIMXRT1170-EVAL.  Did you buy a Segger adapter to connect the JTAG ?  Is yours the J-Link EDU mini ??

**Reply by Thinbungee** [reply]
Hi Robert,  I have a Segger Edu Mini connected to my RT1176 Eval board OK, and can even debug the code sometimes.  What doesnt seem to work yet is loading the code through the debugger.  I recall you set DEBUG=1 in your GNUmakefile.  Can you tell me what the purpose of that setting was ?

**Comment by robert-hh** [comment]
I use a Segger J-Link EDU, but the mini should word as well, besides the plug size. It was little bit tricky to attach. I'm not at my desk atm, but later I can send you a sample script.

**Reply by robert-hh** [reply]
Here is the jdebug file which I use with Ozone. You have to adapt the path names to what you have on your system. Even with that, after loading the code you have to do a manual reset in Ozone by pushing F4. Then the debug starts at main(). I hope it helps.

[mimxrt_m1170.zip](https://github.com/micropython/micropython/files/11044106/mimxrt_m1170.zip)

**Reply by Thinbungee** [reply]
Brilliant ! Will try it soon.  I need to sort out connecting the EDU mini with its 10 way connector but hopefully can get that working soon.  Many thanks for your help again.

**Reply by Thinbungee** [reply]
I have the Segger EDU mini talking to the processor now which is a useful step forward.  It seems that Ozone is x86 only.  Are you building Micropython on a PC virtual machine ?

**Reply by robert-hh** [reply]
I use a debian Linux machine with a x86 CPU.

**Reply by Thinbungee** [reply]
OK, that fits.  Currently I'm building Micropython on Raspberry Pi ok and can run Ozone on PC ok.  But it is a mess to debug because of the file references.  I imagine I can build on a Linux PC virtual machine to sort this out.

**Reply by robert-hh** [reply]
If JLinkGDBServer is available for RPI, you can as well use gdb or kdbg as debugger. Less comfortable and powerful as Ozone, but it works.

**Reply by Thinbungee** [reply]
Hi Robert, I have managed to do a little development for the RT1176 with the RPi compilation and the Segger debugger.  The RPi has proved unreliable however so am now trying to setup with a x86 box and Ubuntu.  From what you have said this sounds similar to your build configuration.  Currently the mpy-cross make will not run however, producing a "file in wrong format" error.  Do you need to alter the make files in your build setup for the x86 host architecture ?

**Reply by robert-hh** [reply]
I run a x86 computer with debian Linux, and did not have to change anything at the make files. mpy-cross is a host computer application, in that case a linux app.

**Reply by Thinbungee** [reply]
Thanks as always for the fast reply.  I cleared the mpy-cross error by deleting all the .o files and totally rebuilding.  Ozone is working much better on the x86 box with Ubuntu so am really pleased to have a much more efficient development system.  All the best, Chris

**Comment by robert-hh** [comment]
Loading the code through the debugger works here with my both the Segger EDU and the EDU mini. You may have to disconnect the debug USB port.
The option DEBUG=1 during build disables optimization. That disables inlining by the compiler. That makes it easier during debugging to single step through the code.

**Comment by Thinbungee** [comment]
One little thing I have fixed for the MIMXRT1170 EVK board is the LEDs.  They now both work from MicroPython on my board.  Am not sure how best for a newbie amateur to get these changes reviewed and incorporated ??

**Reply by robert-hh** [reply]
You have to create a Pull Request including only the change that you made.
